### PR TITLE
fix(bootstrap): Avoid duplicates in create_topics

### LINKED ISSUE #671: Segmentation Fault when running bootstrap
So I've seen this happen in a couple different scenarios. I'm trying to run Sentry in Kubernetes. After getting everything installed, I go to bootstrap Snuba (create the Kafka topics). I've experienced this with both Confluent Kafka as well as Apache Kafka (and multiple versions of each). I've also experienced this in both Minikube and AWS EKS clusters. 

```
/usr/src/snuba# LOG_LEVEL=debug snuba bootstrap --force
2020-01-08 18:39:41,151 Using Kafka with ('kafka-cp-kafka-headless.kafka:9092',)
2020-01-08 18:39:41,165 Attempting to connect to Kafka (attempt 0)
Segmentation fault (core dumped)
```

But if I add some debug log statements, it starts to work....
Here is my `git diff` which caused it to work suddenly 

```
diff --git a/snuba/cli/bootstrap.py b/snuba/cli/bootstrap.py
index 28f52f8..23a85fb 100644
--- a/snuba/cli/bootstrap.py
+++ b/snuba/cli/bootstrap.py
@@ -35,7 +35,6 @@ def bootstrap(
     if kafka:
         logger.debug("Using Kafka with %r", bootstrap_server)
         from confluent_kafka.admin import AdminClient, NewTopic

         attempts = 0
         while True:
             try:
@@ -58,6 +57,7 @@ def bootstrap(
                 time.sleep(1)

         topics = []
+        logger.debug("Made Connection Successfully")
         for name in DATASET_NAMES:
             dataset = get_dataset(name)
             table_writer = dataset.get_table_writer()
@@ -71,14 +71,14 @@ def bootstrap(
                             replication_factor=topic_spec.replication_factor,
                         )
                     )
-
+        print("Created Topics")
         for topic, future in client.create_topics(topics).items():
             try:
                 future.result()
                 logger.info("Topic %s created", topic)
             except Exception as e:
                 logger.error("Failed to create topic %s", topic, exc_info=e)
-
+        print("Actually created topics now")
     from snuba.clickhouse.native import ClickhousePool

     attempts = 0
```

It started to work after the 3rd log statement was added. 

Has anyone else experienced this?

### PR DESCRIPTION
**Pull request by BYK**
Fixes #671. During `bootstrap` we pass the `cdc` topic twice to the `create_topics` function, causing a silent crash when trying to parse the resultant futures. I wasn't able to find anything that prohibits duplicate topic names on the API but looking at the source code reveals some topic-name based dictionary indexing at places which I assume breaks certain assumptions when we pass the same topic name twice.

This patch uses a dict to avoid passing duplicate topic names and fixes the issue.

### CODE REVIEW

**Comment by mattrobenolt** [comment]
I’m not sure much about this, but should we yell more loudly or actually error if there are duplicates? What if there are two definitions for the same topic name, but one has 1 partition and the other has 32 partitions? It seems pretty less than ideal that we’d nondeterministically choose one of them as the correct version.

**Comment by BYK** [comment]
>What if there are two definitions for the same topic name, but one has 1 partition and the other has 32 partitions? It seems pretty less than ideal that we’d nondeterministically choose one of them as the correct version.

Yeah, this occurred to me too and I assumed we would have a central place where we manage topics to ensure unity and consistency. Although this didn't turn out to be true, looks like these are only used as `default_topic` where we only specify the topic name and rest stays with the default values.

I think this requires a larger refactor, beyond my involvement with Snuba and the solution I proposed solves the immediate problem without a newly introduced pitfall (we already have this issue and it _eventually_ works with unspecified behavior).

**Comment by mattrobenolt** [comment]
Yeah, makes sense to me. I’m not opposed, but would prefer someone from SNS team to slam approval. I’m just spectating here.

File: snuba/cli/bootstrap.py [hunk 1]
```diff
@@ -57,22 +57,28 @@ def bootstrap(
                     raise
                 time.sleep(1)
 
-        topics = []
+        topics = dict()
```
Comment (tkaemming) [inline snuba/cli/bootstrap.py]:
```suggestion
        topics = {}
```

File: snuba/cli/bootstrap.py [hunk 2]
```diff
@@ -57,22 +57,28 @@ def bootstrap(
...
+                    topics[topic_spec.topic_name] = NewTopic(
+                        topic_spec.topic_name,
+                        num_partitions=topic_spec.partitions_number,
+                        replication_factor=topic_spec.replication_factor,
                     )
 
-        for topic, future in client.create_topics(topics).items():
+        logger.debug("Initiating topic creation")
+        for topic, future in client.create_topics(
+            list(topics.values()), operation_timeout=1
```
Comment (tkaemming) [inline snuba/cli/bootstrap.py]: I don't have any specific reason to think this, but I wouldn't be surprised if this timeout is too low.

**Review by tkaemming — APPROVED** [review]
I agree that defining topics with conflicting configuration is an issue, I also agree that it's out of scope for this change.

File: snuba/cli/bootstrap.py (hunk 2 above)
Comment (tkaemming) [reply snuba/cli/bootstrap.py]: Actually, maybe not. I was just testing something else out with this `operation_timeout` argument locally — without a broker running, setting this to any value didn't prevent `future.result` from blocking indefinitely. `request_timeout` does work here when the broker can't be reached.

Comment (BYK) [reply snuba/cli/bootstrap.py]: >without a broker running, setting this to any value didn't prevent future.result from blocking indefinitely. request_timeout does work here when the broker can't be reached

I think the previous loop is enough to make sure we have a broker reachable but I'm okay removing that part and baking the retry/request timeout logic into this part (which means we'll have this part in a loop too).

If you think this is a good idea, I'll submit a follow-up patch.

Comment (tkaemming) [reply snuba/cli/bootstrap.py]: > I think the previous loop is enough to make sure we have a broker reachable

Oh right, good point. I think this is fine as-is.

**Pull request merged**